cabal-install/src/Distribution/Client/Init/Types.hs

```diff
@@ -282,13 +289,23 @@ mkLiterate _ hs = hs
 -- -------------------------------------------------------------------- --
 -- Interactive prompt monad
 
+type PromptIO = ReaderT (Data.IORef.IORef SessionState) IO
+
+runPromptIO :: PromptIO a -> IO a
+runPromptIO pio =
+  (Data.IORef.newIORef _newSessionState) >>= (runReaderT pio)
+
 newtype PurePrompt a = PurePrompt
-  { _runPrompt
-      :: NonEmpty String
-      -> Either BreakException (a, NonEmpty String)
+  { _runPromptState
+      :: (NonEmpty String, SessionState)
```

Comment:
This type becomes more and more unpleasant to read. Perhaps you could take a moment and put the NonEmpty String behind a synonym like `Inputs` or `Tape` or some such...